File: Lua/implementação_computacional_1.lua
-- Biblioteca plotly realza a plotagem gráfica das soluções
Plotly = require("plotly")
Grafico = Plotly.figure()

-- Ponto de partida do PVI
local x0 = 0.0
local y0 = 1.0

-- Passo h no eixo x 
H = 0.1

-- Limite superior do eixo x
Limite = 5.0

-- Lista (tabela índice-valor) com os valores de x que são pares ordenados das soluções 
X = {}

-- Preenchimento da lista X
for Valor = x0, Limite, H 
do
    table.insert(X, Valor)
end

-- Cada Y representa uma lista com os resultados de cada método de resolução indicado
Y_Analitico = {}
Y_Euler = {}
Y_Runge_Kutta = {}

-- retorna o valor da derivada y' = f(x, y) = -y³
function f(x, y)
    local resultado = -(y * y * y)
    return resultado
end

-- Analítico
Y_Analitico[1] = y0
for n = 2, #X, 1
do
    Y_Analitico[n] = 1/(math.sqrt(2 * X[n] + 1))
end

-- Euler
Y_Euler[1] = y0
for n = 2, #X, 1
do
    Y_Euler[n] = Y_Euler[n - 1] + H * f(X[n - 1], Y_Euler[n - 1])
end

-- Runge-Kutta
Y_Runge_Kutta[1] = y0
for n = 2, #X, 1
do
    local Kn1 = f(X[n - 1], Y_Runge_Kutta[n - 1])
    local Kn2 = f(X[n - 1] + 1/2 * H, Y_Runge_Kutta[n - 1] + 1/2 * H * Kn1)
    local Kn3 = f(X[n - 1] + 1/2 * H, Y_Runge_Kutta[n - 1] + 1/2 * H * Kn2)
    local Kn4 = f(X[n - 1] + H, Y_Runge_Kutta[n - 1] + H * Kn3)

    Y_Runge_Kutta[n] = Y_Runge_Kutta[n - 1] + H * 1/6 * (Kn1 + 2*Kn2 + 2*Kn3 + Kn4)
end

-- Plotagem do gráfico
Grafico:plot{X, Y_Analitico, mode="lines+markers", marker={symbol="circle", color="green"}, ls="-", color="green", name="Solução Real"}
Grafico:plot{X, Y_Euler, mode="lines+markers", marker={symbol="circle", color="red"}, ls="--", color="red", name="Método de Euler"}
Grafico:plot{X, Y_Runge_Kutta, mode="lines+markers", marker={symbol="circle", color="blue"}, ls="--", color="blue", name="Método de Runge-Kutta"}

Grafico:plot{title="Gráfico com as resoluções do PVI {y' = -y³; y(0) = 1}, com h = 0.1 e 0 <= x <= 5"}
Grafico:plot{xlabel="Eixo x"}
Grafico:plot{ylabel="Eixo y"}
Grafico:plot{legend="True"}
Grafico:plot{grid="True"}:show()

File: Python/implementação_computacional_1.py
# Resolução do PVI {y' = -y³; y(0) = 1} usando os métodos de Euler e Runge-Kutta.
# Gráfico apresentando a diferença de resultado entre o método analítico, de Euler e de Runge-Kutta.

# Import da biblioteca de matemática e da biblioteca de plotagen
import math
import matplotlib.pyplot as grafico

# Ponto fornecido pelo PVI
x0 = 0.0
y0 = 1.0

# Limite do intervalo (0 <= x <= 5)
limite = 5.0

# Passo h
h = 0.1

# Valores do eixo x estarão contidos na lista X
X = []

# Preenchimento de X
valor = x0
while valor <= limite:
    X.append(valor)
    valor += h


# Cada Y representa uma lista com os resultados de cada método de resolução indicado
# Inicialmente são preenchidas com 0 em todas as posições para evitar erros de indície em atribuições futuras
# len(X) é o tamanho da lista X já preenchida, as listas Y devem ter o mesmo tamanho de X para que a resolução esteja correta
Y_Analitico = [0] * len(X)
Y_Euler = [0] * len(X)
Y_Runge_Kutta = [0] * len(X)


# Derivada y' = f(x,y) = -y³
def f(x, y):
    resultado = -(y ** 3)
    return resultado


# Resolução analítica, Y_Analitico começa recebendo y0 na primeira posição graças ao PVI
Y_Analitico[0] = y0
n = 1
while n <= len(X) - 1:
    Y_Analitico[n] = 1/(math.sqrt(2 * X[n] + 1))
    n += 1


# Resolução pelo método de Euler, Y_Euler começa recebendo y0 na primeira posição por definição
# y(n) = y(n-1) + h * f(Xn-1, Yn-1)
Y_Euler[0] = y0
n = 1
while n <= len(X) - 1:
    Y_Euler[n] = Y_Euler[n - 1] + h * f(X[n - 1], Y_Euler[n - 1])
    n += 1


# Resolução pelo método de Runge-Kutta, Y_Runge_Kutta começa recebendo y0 na primeira posição por definição
# y(n) = y(n-1) + 1/6 * h * (Kn1 + 2Kn2 + 2Kn3 + Kn4)
Y_Runge_Kutta[0] = y0
n = 1
while n <= len(X) - 1:
    Kn1 = f(X[n - 1], Y_Runge_Kutta[n - 1])
    Kn2 = f(X[n - 1] + 1/2 * h, Y_Runge_Kutta[n - 1] + 1/2 * h * Kn1)
    Kn3 = f(X[n - 1] + 1/2 * h, Y_Runge_Kutta[n - 1] + 1/2 * h * Kn2)
    Kn4 = f(X[n - 1] + h, Y_Runge_Kutta[n - 1] + h * Kn3)

    Y_Runge_Kutta[n] = Y_Runge_Kutta[n - 1] + h * 1/6 * (Kn1 + 2*Kn2 + 2*Kn3 + Kn4)
    n += 1


# Plotagem do gráfico com os resultados dos três métodos
grafico.plot(X, Y_Runge_Kutta, marker='o', linestyle='--', color='b', label='Método de Runge-Kutta')
grafico.plot(X, Y_Euler, marker='o', linestyle='--', color='r', label='Método de Euler')
grafico.plot(X, Y_Analitico, marker='o', linestyle='-', color='g', label='Solução Exata')

grafico.title('Gráfico com as resoluções do PVI {y\' = -y³; y(0) = 1}, com h = 0.1 e 0 <= x <= 5')
grafico.xlabel('Eixo x')
grafico.ylabel('Eixo y')

grafico.grid(True)
grafico.legend()
grafico.show()